### PR TITLE
Implementing Plugins for EventHub

### PR DESCRIPTION
**Pull request by davidrevoledo**
## Description
This is an Implementation for Plugins supported in the NetClient Event Hub
The approach is support the same feature ServiceBus does, which is handy.

The code doesn't have include unit test, if the EventHub team are interested in support this feature I can go ahead and complete the required UT.

I think it'd be useful, I've to use it and have to do some workarounds to  overwrite `EventDataSender`

This checklist is used to make sure that common guidelines for a pull request are followed.

- [X] **I have read the [contribution guidelines](./CONTRIBUTING.md).**
- [X] Title of the pull request is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR.
- [X] The pull request does not introduce breaking changes (unless a major version change occurs in the assembly and module).
- [X] If applicable, the public code is properly documented.
- [x] Pull request includes test coverage for the included changes.
- [X] The code builds without any errors.

### CODE REVIEW

**Comment by serkantkaraca** [comment]
Thanks David, I will review the proposed changes. I wonder if I misread your first sentence. The plugin model isn't supported .Net EH library today. Are you referring to some other library?

**Comment by davidrevoledo** [comment]
@serkantkaraca Yes the same feature is supported by Service bus dotnet libarary, which is handy, because you can dynamically set plugins to the clients, and decouple implementations.  event differents plugins for differents tenancies, I'm using a lot this feature in Service Bus, and currently doing some workarounds to do the same with Event bus.

**Comment by serkantkaraca** [comment]
OK, then we are on the same page. I thought you were referring to https://www.nuget.org/packages/WindowsAzure.ServiceBus

**Comment by davidrevoledo** [comment]
@serkantkaraca yeah the same (one is the net framework the other the new one with foundation) : )

File: src/Microsoft.Azure.EventHubs/EventDataSender.cs [hunk 1]
```diff
@@ -40,6 +44,50 @@ internal static int ValidateEvents(IEnumerable<EventData> eventDatas)
             return count;
         }
 
+        async Task<EventData> ProcessEvent(EventData @event)
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs/EventDataSender.cs]: >@event [](start = 53, length = 6)

Let's use eventData here.

File: src/Microsoft.Azure.EventHubs/EventHubClient.cs [hunk 2]
```diff
@@ -1,6 +1,9 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System.Linq;
+using Microsoft.Azure.EventHubs.Core;
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs/EventHubClient.cs]: Move under namespace

File: src/Microsoft.Azure.EventHubs/EventHubsEventSource.cs [hunk 3]
```diff
@@ -1,6 +1,8 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System;
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs/EventHubsEventSource.cs]: Move under namespace

File: src/Microsoft.Azure.EventHubs/EventHubsEventSource.cs [hunk 4]
```diff
@@ -209,6 +211,33 @@ public void ReceiveHandlerExitingWithError(string clientId, string partitionId,
             }
         }
 
+        [Event(22, Level = EventLevel.Verbose, Message = "User plugin {0} called on client {1}")]
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs/EventHubsEventSource.cs]: >22 [](start = 15, length = 2)

Let's start plugin events from 100

File: src/Microsoft.Azure.EventHubs/EventDataSender.cs (hunk 1 above)
Comment (davidrevoledo) [reply src/Microsoft.Azure.EventHubs/EventDataSender.cs]: Perfect will check if in other place there is a @event also

File: src/Microsoft.Azure.EventHubs/Resources.Designer.cs [hunk 5]
```diff
@@ -39,7 +39,7 @@ internal class Resources {
         internal static global::System.Resources.ResourceManager ResourceManager {
             get {
                 if (object.ReferenceEquals(resourceMan, null)) {
-                    global::System.Resources.ResourceManager temp = new global::System.Resources.ResourceManager("Microsoft.Azure.EventHubs.Resources", typeof(Resources).GetTypeInfo().Assembly);
+                    global::System.Resources.ResourceManager temp = new global::System.Resources.ResourceManager("Microsoft.Azure.EventHubs.Resources", typeof(Resources).Assembly);
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs/Resources.Designer.cs]: This change will probably break one of the builds. You can revert the change manually if needed.

File: src/Microsoft.Azure.EventHubs/Core/EventHubPlugin.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs/Core/EventHubPlugin.cs]: Move under namespace

File: src/Microsoft.Azure.EventHubs/Core/EventHubPlugin.cs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+namespace Microsoft.Azure.EventHubs.Core
+{
+    /// <summary>
+    ///     This class provides methods that can be overridden to manipulate messages for custom plugin functionality.
+    /// </summary>
+    public abstract class EventHubPlugin
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs/Core/EventHubPlugin.cs]: >EventHub [](start = 26, length = 8)

EventHubs

File: src/Microsoft.Azure.EventHubs/Core/EventHubPlugin.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        /// <remarks>This name is used to identify the plugin, and prevent a plugin from being registered multiple times.</remarks>
+        public abstract string Name { get; }
+
+        /// <summary>
+        ///     Determines whether or an exception in the plugin should prevent a send or receive operation.
+        /// </summary>
+        public virtual bool ShouldContinueOnException => false;
+
+        /// <summary>
+        ///     This operation is called before a event is sent.
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs/Core/EventHubPlugin.cs]: >a [](start = 48, length = 1)

an

File: src/Microsoft.Azure.EventHubs/Core/EventHubPlugin.cs [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        /// </summary>
+        /// <param name="eventData">The <see cref="EventData" /> to be modified by the plugin</param>
+        /// <returns>The modified event <see cref="EventData" /></returns>
+        public virtual Task<EventData> BeforeMessageSend(EventData eventData)
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs/Core/EventHubPlugin.cs]: >BeforeMessageSend [](start = 39, length = 17)

Makes sense to have AfterMessageSend?

Comment (davidrevoledo) [reply src/Microsoft.Azure.EventHubs/Core/EventHubPlugin.cs]: Yes for that I've to change the processor I can do, just want to be sure that the Feature was not a conflict with the roadmap to go ahead.

File: src/Microsoft.Azure.EventHubs/Core/EventHubPlugin.cs (hunk 8 above)
Comment (davidrevoledo) [reply src/Microsoft.Azure.EventHubs/Core/EventHubPlugin.cs]: thanks

File: src/Microsoft.Azure.EventHubs/Core/EventHubPlugin.cs (hunk 7 above)
Comment (davidrevoledo) [reply src/Microsoft.Azure.EventHubs/Core/EventHubPlugin.cs]: thanks

File: src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs [hunk 10]
```diff
@@ -1,6 +1,10 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Azure.EventHubs.Core;
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs]: Move under namespace

File: src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs [hunk 11]
```diff
@@ -55,6 +63,43 @@ public RetryPolicy RetryPolicy
         /// <returns>The asynchronous operation</returns>
         public abstract Task CloseAsync();
 
+        /// <summary>
+        /// Registers a <see cref="EventHubPlugin"/> to be used with this sender.
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs]: >sender [](start = 74, length = 6)

This may not be a sender. Let's keep this generic.

File: src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs [hunk 12]
```diff
@@ -55,6 +63,43 @@ public RetryPolicy RetryPolicy
...
+        /// </summary>
+        public virtual void RegisterPlugin(EventHubPlugin eventHubPlugin)
+        {
+            if (eventHubPlugin == null)
+            {
+                throw new ArgumentNullException(nameof(eventHubPlugin), Resources.ArgumentNullOrWhiteSpace.FormatForUser(nameof(eventHubPlugin)));
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs]: >nameof [](start = 48, length = 6)

Plugin has property 'Name'. Shouldn't  we use it instead for duplicate check?

File: src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs [hunk 13]
```diff
@@ -55,6 +63,43 @@ public RetryPolicy RetryPolicy
...
+            {
+                return;
+            }
+            if (string.IsNullOrWhiteSpace(eventHubPlugin))
+            {
+                throw new ArgumentNullException(nameof(eventHubPlugin), Resources.ArgumentNullOrWhiteSpace.FormatForUser(nameof(eventHubPlugin)));
+            }
+            if (this.RegisteredPlugins.Any(p => p.Name == eventHubPlugin))
+            {
+                var plugin = this.RegisteredPlugins.First(p => p.Name == eventHubPlugin);
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs]: >First [](start = 52, length = 5)

Remove all matching instead of first here in case RegisteredPlugins has duplicates.

File: src/Microsoft.Azure.EventHubs/EventDataSender.cs (hunk 1 above)
Comment (davidrevoledo) [reply src/Microsoft.Azure.EventHubs/EventDataSender.cs]: Also changed here https://github.com/Azure/azure-event-hubs-dotnet/pull/324/files#diff-a221f37e8c378f551b70067d20e88d6eR82

File: src/Microsoft.Azure.EventHubs/EventHubClient.cs [hunk 14]
```diff
@@ -10,6 +10,8 @@ namespace Microsoft.Azure.EventHubs
     using System.Threading.Tasks;
     using Microsoft.Azure.EventHubs.Amqp;
     using Microsoft.IdentityModel.Clients.ActiveDirectory;
+    using System.Linq;
+    using Microsoft.Azure.EventHubs.Core;
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs/EventHubClient.cs]: Can you sort usings? Same for the other classes.

**Comment by davidrevoledo** [comment]
@serkantkaraca Hey Serkant  I made the proposed changes.

However didn't get that 
https://github.com/Azure/azure-event-hubs-dotnet/pull/324#discussion_r223782942
`Remove all matching instead of first here in case RegisteredPlugins has duplicates.`

if a plugin with same name is configured this will break no take the last one so the first will break the code there can not be more than 1 with the same name, I changed the error message to take the plugin name instead for more clarification.

File: src/Microsoft.Azure.EventHubs/EventHubClient.cs (hunk 14 above)
Comment (davidrevoledo) [reply src/Microsoft.Azure.EventHubs/EventHubClient.cs]: Sure ! :)

**Comment by serkantkaraca** [comment]
RegisteredPlugins is a generic List. It can end up with duplicates easily due to race. If you think having duplicates will hurt then you can try using ConcurrentDictionary or a lock. IMHO, duplicates won't hurt but let's do the cleanup accounting duplicate potential here.

---
In reply to: [428391883](https://github.com/Azure/azure-event-hubs-dotnet/pull/324#issuecomment-428391883) [](ancestors = 428391883)

**Comment by serkantkaraca** [comment]
> {

Can you sort all usings and also use full namespace? Same for other classes.

---
Refers to: src/Microsoft.Azure.EventHubs/EventHubClient.cs:5 in 3ecbaac. [](commit_id = 3ecbaac9d66a711a154cf437c5fd9de522f33d25, deletion_comment = False)

File: src/Microsoft.Azure.EventHubs.Processor/EventHubPartitionPump.cs [hunk 15]
```diff
@@ -140,14 +140,36 @@ public PartitionReceiveHandler(EventHubPartitionPump eventHubPartitionPump)
 
             public int MaxBatchSize { get; set; }
 
-            public Task ProcessEventsAsync(IEnumerable<EventData> events)
+            public async Task ProcessEventsAsync(IEnumerable<EventData> events)
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs.Processor/EventHubPartitionPump.cs]: Can you remove the plugin from processor? I don't really see a use case here. We can add it later if there is any ask from customers.

Comment (davidrevoledo) [reply src/Microsoft.Azure.EventHubs.Processor/EventHubPartitionPump.cs]: Sure for now I'm only using for Sender but After Receive then won't be used shall I remove this right ? I added that because this https://github.com/Azure/azure-event-hubs-dotnet/pull/324#discussion_r223779579

Comment (serkantkaraca) [reply src/Microsoft.Azure.EventHubs.Processor/EventHubPartitionPump.cs]: Actually that comment is for AfterMessageSend, not receive. Let's do BeforeMessageSend for now and see how that rolls out.

**Comment by davidrevoledo** [comment]
@serkantkaraca https://github.com/Azure/azure-event-hubs-dotnet/pull/324#issuecomment-428642829
Reagarding this, really up to you, I think it can be a quite confusing if you don't know they are being replaced and for consistency with the rest of ServiceBus packages I'd use the exception.

I think same as you duplicated don't hurt, but in the context it can be really confusing. 

ConcurrencyDictionary suit perfect IMO.

Waiting for your feedback to do the changes thought :) 

Great thanks @serkantkaraca

File: src/Microsoft.Azure.EventHubs.Processor/EventHubPartitionPump.cs (hunk 15 above)
Comment (davidrevoledo) [reply src/Microsoft.Azure.EventHubs.Processor/EventHubPartitionPump.cs]: Perfect yes I don't think an scenario to use `AfterMessageSend`, for Service bus probably yes because ServiceBus write for you MessageId property if you didn't.

Comment (serkantkaraca) [reply src/Microsoft.Azure.EventHubs.Processor/EventHubPartitionPump.cs]: Fair enough. let's keep it as BeforeSend for now.

**Comment by serkantkaraca** [comment]
Let's do ConcurrentDictionary then. Use TryAdd and throw 'already registered' if returns false. It also makes sense to use plugin name as key of the entry.

[https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.tryadd?redirectedfrom=MSDN&view=netframework-4.7.2#System_Collections_Concurrent_ConcurrentDictionary_2_TryAdd__0__1](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.tryadd?redirectedfrom=MSDN&view=netframework-4.7.2#System_Collections_Concurrent_ConcurrentDictionary_2_TryAdd__0__1)_

---
In reply to: [428670630](https://github.com/Azure/azure-event-hubs-dotnet/pull/324#issuecomment-428670630) [](ancestors = 428670630)

**Comment by davidrevoledo** [comment]
@serkantkaraca Do you think shall the package throw an error if TryDelete return false removing the plugin also what do you think about messages 
` There was an error trying to add the {0} plugin` and if apply 

`There was an error trying to remove the {0} plugin`  ?

**Comment by serkantkaraca** [comment]
Let's not throw in that case.

**Comment by davidrevoledo** [comment]
@serkantkaraca I was thinking the same :)

**Comment by davidrevoledo** [comment]
@serkantkaraca Done :)

File: src/Microsoft.Azure.EventHubs/Resources.Designer.cs (hunk 5 above)
Comment (davidrevoledo) [reply src/Microsoft.Azure.EventHubs/Resources.Designer.cs]: @serkantkaraca  Why is the reason this fail w/o `typeof(Resources).GetTypeInfo().Assembly`?

**Comment by serkantkaraca** [comment]
> {

Sorry for misunderstanding. The order before change actually is correct. First System then the rest.

---
In reply to: [428643337](https://github.com/Azure/azure-event-hubs-dotnet/pull/324#issuecomment-428643337) [](ancestors = 428643337)

---
Refers to: src/Microsoft.Azure.EventHubs/EventHubClient.cs:5 in 3ecbaac. [](commit_id = 3ecbaac9d66a711a154cf437c5fd9de522f33d25, deletion_comment = False)

File: src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs [hunk 16]
```diff
@@ -55,6 +65,43 @@ public RetryPolicy RetryPolicy
         /// <returns>The asynchronous operation</returns>
         public abstract Task CloseAsync();
 
+        /// <summary>
+        /// Registers a <see cref="EventHubsPlugin"/> to be used with this client.
+        /// </summary>
+        public virtual void RegisterPlugin(EventHubsPlugin eventHubPlugin)
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs]: >eventHub [](start = 59, length = 8)

Can you make sure it is 'eventhubs' everywhere since this needs to be same as the service name. 'Event Hub' refers to single entity.

File: src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs [hunk 17]
```diff
@@ -55,6 +65,43 @@ public RetryPolicy RetryPolicy
...
+            {
+                throw new ArgumentException(nameof(eventHubPlugin), Resources.PluginRegistrationFailed.FormatForUser(nameof(eventHubPlugin.Name)));
+            }
+        }
+
+        /// <summary>
+        /// Unregisters a <see cref="EventHubsPlugin"/>.
+        /// </summary>
+        /// <param name="eventHubPlugin">The <see cref="EventHubsPlugin.Name"/> of the plugin to be unregistered.</param>
+        public virtual void UnregisterPlugin(string eventHubPlugin)
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs]: >eventHubPlugin [](start = 52, length = 14)

You can use 'pluginName' here I guess.

File: src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs [hunk 18]
```diff
@@ -55,6 +65,43 @@ public RetryPolicy RetryPolicy
...
+        public virtual void RegisterPlugin(EventHubsPlugin eventHubPlugin)
+        {
+            if (eventHubPlugin == null)
+            {
+                throw new ArgumentNullException(nameof(eventHubPlugin), Resources.ArgumentNullOrWhiteSpace.FormatForUser(nameof(eventHubPlugin)));
+            }
+            if (this.RegisteredPlugins.Any(p => p.Key == eventHubPlugin.Name))
+            {
+                throw new ArgumentException(nameof(eventHubPlugin), Resources.PluginAlreadyRegistered.FormatForUser(nameof(eventHubPlugin.Name)));
+            }
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs]: You don't need this anymore since TryAdd does same job.

File: src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs [hunk 19]
```diff
@@ -55,6 +65,43 @@ public RetryPolicy RetryPolicy
...
+        /// </summary>
+        public virtual void RegisterPlugin(EventHubsPlugin eventHubPlugin)
+        {
+            if (eventHubPlugin == null)
+            {
+                throw new ArgumentNullException(nameof(eventHubPlugin), Resources.ArgumentNullOrWhiteSpace.FormatForUser(nameof(eventHubPlugin)));
+            }
+            if (this.RegisteredPlugins.Any(p => p.Key == eventHubPlugin.Name))
+            {
+                throw new ArgumentException(nameof(eventHubPlugin), Resources.PluginAlreadyRegistered.FormatForUser(nameof(eventHubPlugin.Name)));
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs]: >nameof [](start = 116, length = 6)

just plugin.Name, right?

File: src/Microsoft.Azure.EventHubs/Resources.Designer.cs (hunk 5 above)
Comment (serkantkaraca) [reply src/Microsoft.Azure.EventHubs/Resources.Designer.cs]: I don't really remember the exact reason but I guess some assembly binding issue.

---
In reply to: [224279162](https://github.com/Azure/azure-event-hubs-dotnet/pull/324#discussion_r224279162) [](ancestors = 224279162)

File: src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs (hunk 17 above)
Comment (davidrevoledo) [reply src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs]: Sure

File: src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs (hunk 18 above)
Comment (davidrevoledo) [reply src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs]: Actually according this  : https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.tryadd?redirectedfrom=MSDN&view=netframework-4.7.2#System_Collections_Concurrent_ConcurrentDictionary_2_TryAdd__0__1

It throws `ArgumentNullException` Only if key is null which will never be the case because above is this code 

C# ```
            if (string.IsNullOrWhiteSpace(pluginName))
            {
                throw new ArgumentNullException(nameof(pluginName), Resources.ArgumentNullOrWhiteSpace.FormatForUser(nameof(pluginName)));
            }
```

Just my opinion but it's handy get a bit more information there about which one failed also.

File: src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs (hunk 19 above)
Comment (davidrevoledo) [reply src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs]: You mean this instead ? 

```C#
if (this.RegisteredPlugins.Any(p => p.Value.Name == eventHubPlugin.Name))
            {
                throw new ArgumentException(nameof(eventHubPlugin), Resources.PluginAlreadyRegistered.FormatForUser(nameof(eventHubPlugin.Name)));
            }
```

Comment (serkantkaraca) [reply src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs]: nameof will return 'Name' and the error message will build as: The Name plugin has already been registered.

---
In reply to: [224288914](https://github.com/Azure/azure-event-hubs-dotnet/pull/324#discussion_r224288914) [](ancestors = 224288914)

Comment (davidrevoledo) [reply src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs]: Fair enought

**Comment by davidrevoledo** [comment]
@serkantkaraca done !

**Comment by serkantkaraca** [comment]
Seems this is not fixed yet.

nameof(eventHubsPlugin.Name)

**Comment by davidrevoledo** [comment]
@serkantkaraca  which Line  ?  I've changed here 

https://github.com/Azure/azure-event-hubs-dotnet/pull/324/commits/0f54c6b7fab2dea2f334d45072a7ec44d8d0d231#diff-95e8d87b5786ff706c6992e246351d0bR82

**Comment by serkantkaraca** [comment]
Same here:

throw new ArgumentNullException(nameof(pluginName), Resources.ArgumentNullOrWhiteSpace.FormatForUser(nameof(pluginName)));

**Comment by davidrevoledo** [comment]
Sure so sorry because the misunderstanding @serkantkaraca

**Comment by davidrevoledo** [comment]
@serkantkaraca Ok sorry because the misunderstanding 

I can change to this
image

Can't put nameof(eventHubsPlugin.Name) either just eventHubsPlugin.Name as the object is null and will crash.

Are you agree with this ?

**Comment by serkantkaraca** [comment]
My comment is for lines 78 and 82 only.

**Comment by davidrevoledo** [comment]
@serkantkaraca Ok thanks

File: src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs [hunk 20]
```diff
@@ -1,11 +1,15 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System.Collections.Concurrent;
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs]: Move under namespace section

**Comment by serkantkaraca** [comment]
Added one more comment and after that I think we are all good. Thx!

File: src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs (hunk 20 above)
Comment (davidrevoledo) [reply src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs]: Ohh resharper move it automatically and I didn't aware of that, super sorry will configure to avoid that in a future I did with other configurations :) .

**Comment by davidrevoledo** [comment]
@serkantkaraca No Serkat thanks to you for let me participate and for you patient in this big chunk of code, please if anytime you need help because of time or need something to investigate is needed feel free to ask :D will be more than happy to help. !

**Comment by serkantkaraca** [comment]
Seems below test is failing. Can you take a look?

Plugin_without_ShouldContinueOnException_should_throw

**Comment by davidrevoledo** [comment]
One last thing to solve but then celebrate the feature 🎉
